service/history/shard/controller_impl.go

```diff
@@ -446,10 +481,67 @@ func (c *ControllerImpl) acquireShards(ctx context.Context) {
 	_ = sem.Acquire(ctx, concurrency)
 
 	c.RLock()
+	// note that this count includes lingering shards
 	numOfOwnedShards := len(c.historyShards)
 	c.RUnlock()
 	metrics.NumShardsGauge.With(c.taggedMetricsHandler).Record(float64(numOfOwnedShards))
 	c.publishShardCountUpdate(numOfOwnedShards)
+
+	// Readiness check: We should set initialShardsAcquired when:
+	// 1. It's not already set.
+	// 2. We should own at least one shard (i.e. not before we join membership).
+	// 3. We have ownership of all the shards we're supposed to own.
+	if readinessCtx != nil {
+		if len(ownedShards) > 0 {
+			go func() {
+				defer readinessCancel()
+				if c.checkShardReadiness(readinessCtx, ownedShards) {
+					c.initialShardsAcquired.SetIfNotReady(struct{}{}, nil)
+				}
+			}()
+		} else {
+			readinessCancel()
+		}
+	}
+}
+
+func (c *ControllerImpl) checkShardReadiness(
+	ctx context.Context,
+	shards []int32,
+) bool {
+	concurrency := int64(max(c.config.AcquireShardConcurrency(), 1))
+	sem := semaphore.NewWeighted(concurrency)
+	var ready atomic.Int32
+	for _, shardID := range shards {
+		if sem.Acquire(ctx, 1) != nil {
+			return false
+		}
+		go func() {
+			defer sem.Release(1)
+			// Note that AssertOwnership uses a detached context for the actual persistence
+			// op so we can't cancel it. If context is canceled, the final Acquire will
+			// fail and we won't do anything.
+			if shard, err := c.GetShardByID(shardID); err != nil {
+				return
+			} else if _, err := shard.GetEngine(ctx); err != nil {
+				return
+			} else if shard.AssertOwnership(ctx) != nil {
+				return
+			}
```

Comment:
are any of those expected? if not, shall we wrote log with enough debug info, or emit metrics etc?

Reply:
`acquireShards` already has lots of metrics and some logs. The main observable result of this readiness check is the grpc health check status. I don't think it needs any other logs or metrics itself (though there is one log line for the whole thing below)